File: 백준/Silver/1912. 연속합/연속합.py
import sys
"""
dp 배열 내부에서 최대값 찾는 방식으로 해결하려다 실패함
dp 배열 내부의 최대값을 리턴하지 말고, 기억하는 내용을 리셋하고 지금까지 최대값 더하는 식으로 접근 변경함
"""


sin = sys.stdin.readline
N = int(sin())
numbers = list(map(int, sin().split(" ")))
dp = [-1001 for _ in range(N)]
max_value = numbers[0]
dp[0] = numbers[0]

for i in range(1, N):
    dp[i] = max(dp[i-1] + numbers[i], numbers[i])
    max_value = max(dp[i], max_value)

# print(dp)
print(max_value)

File: 백준/Silver/1912. 연속합/연속합.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        int N = Integer.parseInt(br.readLine());
        int[] numbers = new int[N];
        int[] dp = new int[N];

        String[] ipts = br.readLine().split(" ");

        for (int i = 0; i < N; i++) {
            numbers[i] = Integer.parseInt(ipts[i]);
            dp[i] = numbers[i];
        }

        int answer = dp[0];

        for (int i = 1; i < N; i++) {
            dp[i] = Math.max(dp[i], dp[i-1] + numbers[i]);
            answer = Math.max(dp[i], answer);
        }

        System.out.println(answer);
    }
}
